titan/features/partner_tracing.py

```diff
@@ -21,35 +62,167 @@ def update_agent(self, model: "model.TITAN"):
-                and not agent_exposure.dx
-                and model.run_random.random() < params.dx_prob
+                agent_exposure.dx
+                and agent_exposure.dx_time < model.time
+                and not self.ps_prev_dx # and not already marked
             ):
-                agent_exposure.diagnose(model)
-
+                self.ps_prev_dx = True
+                self.ps_prev_dx_time = model.time
+                # is this agent on haart? if not, mark
+                if not self.agent.haart.active:
+                    self.prev_dx_not_on_haart = True
+                    
+            # if this traced agent has not been diagnosed, test agent
+            if not agent_exposure.dx:
+                self.test_agent(agent_exposure, params, model)
+                    
+                    
+            # if agent was traced via PS and was previously diagnosed with hiv and is not on haart, stochastically enroll in haart
+            if (
+                  agent_exposure.dx
+                  and agent_exposure.dx_time < model.time
+                  and not self.agent.haart.active
+                  and model.run_random.random() < params.re_haart_prob
+            ):
+                haart_params = (
+                    self.agent.location.params.demographics[self.agent.race]
+                    .sex_type[self.agent.sex_type]
+                    .drug_type[self.agent.drug_type]
+                    .haart
+                )
+                self.agent.haart.initiate(model.run_random, haart_params, "prob") #deterministic. They will enroll in haart.
+                self.re_initiated = True
+                self.re_initiated_time = model.time
+                
+  
         # stop tracing of this agent if time
         if self.active and model.time >= self.time + params.trace_duration:
             self.active = False
             self.time = None
+        
+        # this is to ensure agents are able to test negative multiple times throughout the simulation if contacted. On the otherhand, agents cannot test positive multiple times
+        if self.tested_negative and model.time > self.tested_negative_time:
+            self.tested_negative = False
+            self.tested_negative_time = None
+        
+        # this is to ensure agents are able to participate multiple times throughout the simulation 
+        if self.ps_participant and model.time > self.ps_participant_time:
+            self.ps_participant = False
+            self.ps_participant_time = None
+    
+    def trace_partners(self, params, model: "model.TITAN"):
+        for ptnr in self.agent.get_partners(params.bond_type):
+            
+            # step 3
+            self.info = True if model.run_random.random() < params.tracing_prob else False # can only trace if participant provides information
+            # step 4
+            self.contact = True if model.run_random.random() < params[ptnr.race].contact_prob else False # contact stochastically
```

Comment:
are self.info and self.contact used other than the if statement below? If not, no need to define them as an attribute.